Hapi patch

AdriVanHoudt opened the pull request:
Update hapi instrumentation by making sure to always set the the transaction name, onPreResponse does not get called when the route errors and OPTIONS calles do not follow the same lifecycle

I tested it on hapi 13 but I do not expect it to fail on hapi9+ since it is pretty generic

(fyi if you want I can add some extra stuff that will automatically log errors decently from hapi, something I do manually now)

AdriVanHoudt commented:
Btw the main driver behind this patch is that calls that error are logged as `unknown route` and OPTIONS as well

watson commented:
@AdriVanHoudt Thank you for the pull request :smiley: Highly appreciated!

Looking it over now

watson commented on lib/instrumentation/modules/hapi.js:
```diff
@@ -9,22 +9,41 @@ module.exports = function (hapi, agent) {
+
+          if (request.method === 'options') {
+            var name = (request.raw && request.raw.req && request.raw.req.method) ||
+                       (request.route.method && request.route.method.toUpperCase())
```
According to [the official documentation](http://hapijs.com/api#request-lifecycle), the `request.route` property is not yet set in the `onRequest` event. But for some reason it's set when I test it, so maybe it's an error in the documentation or maybe I'm reading this wrong.

In any case, I think it would be a good idea to add a guard against it not being set just in case:

```diff
-(request.route.method && request.route.method.toUpperCase())
+(request.route && request.route.method && request.route.method.toUpperCase())
```

AdriVanHoudt replied:
I see, it is being set here (https://github.com/hapijs/hapi/blob/e724897cc9339935d8900fc2c1ef073eef046d9d/lib/request.js#L98) though so maybe we should in general not rely on `request.route`? I mean `request.raw.req.method` will always resolve no?

watson replied:
Yeah `request.raw.req.method` should always work

watson replied:
For future reference, the `request.route` property is technically present during `onRequest`, but the actual route is not yet populated. So it's actually kind of useless at this point.

watson commented:
@AdriVanHoudt Why did you use both the `onRequest` and the `onPreAuth` extension point? Wouldn't one of them be enough or what was the thought behind that?

AdriVanHoudt commented:
@watson when using CORS, hapi will create special OPTIONS routes for you and handle them for you, these calls will never reach `onPreAuth` and will be logged as unknown. So I added the `onRequest` and changed `onPreResponse` to `onPreAuth` since pre response will not be called on errors and it does allow for better tagging the route then `on request`.
We could only use `onRequest` but we lose the `fingerprinting` then (we could copy it ourselves though)

watson commented:
I'm not sure using `request.url.pathname` is a good choice in `onRequest`. In general the issue with `onRequest` is that Hapi hasn't yet figured out which route the request should be routed to, so the route name isn't available here. So if this is the **only** handler called, we might just better off calling it `unknown route` since grouping on the full URL isn't a good default for most users.

Normally `onPreAuth` will be called right after `onRequest`. It will overwrite the transaction name set in `onRequest` with the more generic route or fingerprint name available after the route have been resolved. But as you mention, it will not be called for `OPTIONS` requests.

You also mention that if an error occurs, it will not call `onPreResponse`. I've tried different things, but it always calls my `onPreResponse` handler - no matter if I throw the error or call `reply(new Error('foo'))`. I might just be doing it wrong of cause. Can you show me a snippet that make it skip `onPreResponse`?

watson commented:
These are basically the rules we want to follow:

- If no route matches the request, we want to give the name `<METHOD> unknown route` where `<METHOD>` is the HTTP method used by the request, e.g. `GET unknown route`
- If a route matches the request we want to use the route fingerprint (or name): `<METHOD> <fingerprint>`, e.g. `GET /users/?`

Note that not all 404's are "unknown routes". You might have a `GET /users/{id}` route and a request that accesses `GET /users/unknown-id`. That request should still be named and grouped together with other `GET /users/?` transactions. Opbeat will take note of the status code and mark it accordingly in the UI.

AdriVanHoudt commented:
I get the grouping part stand but to me, personally, it does not provide much information. I rather have it grouped under specific routes but I can do that manually for me I suppose.
As for the `onPreResponse`, turns out I have another listener on the event to transform error codes that replies itself with the transformed error causing the other `onPreResponse` to not be called (hapi prevents infinite loop here). Setting the name `onPreAuth` solves this.

watson commented:
Ok, let's keep the new `onPreAuth` hook then :smiley: 

Can you show me how you set up cors in hapi?

AdriVanHoudt commented:
:+1: 
I have it just set to `true`, see `cors` under http://hapijs.com/api#route-options. I also pass it at server level so it is the default for all routes via http://hapijs.com/api#new-serveroptions under `routes`

watson commented:
@AdriVanHoudt I've found a way to detect when something is a CORS preflight request and I can then mark it as such in the Opbeat interface (with the name `CORS preflight`). As I mentioned above, there unfortunately doesn't seem to be a way to get a proper route name for the `OPTIONS` routes that are autogenerated by hapi when setting `cors: true`. So all CORS preflight requests will therefore be grouped together no matter their actual URL.

I understand that you don't mind that the unique URL's will generate unique transactions in Opbeat, but for most of our users this is not a good default, so I unfortunately can't implement it in that way.

How would you feel about the default being marking all CORS preflight requests as `CORS preflight` in Opbeat?

AdriVanHoudt commented:
@watson if all the options calls are marked as `CORS preflight` that would be very fine! Thanks for the effort!

AdriVanHoudt commented:
you want me to remove the `onRequest` part of the PR?

watson commented:
@AdriVanHoudt No worries... I'm going to add a the CORS detection anyway, so I'll just merge it and create an extra commit with the modification :smiley:

watson commented:
@AdriVanHoudt one thing: could I get you to change the commit message? The current edition is a bit too long. I'd just change it to this (linebreaks important):

> hapi: improve transaction naming
> 
> Update hapi instrumentation by making sure to always set the the
> transaction name, onPreResponse does not get called when the route
> errors and OPTIONS calles do not follow the same lifecycle

AdriVanHoudt commented:
@watson  Ok I changed the message

watson commented:
@AdriVanHoudt It's now :ship:'ed to npm as v3.6.0 :tada:
Thanks a lot for your help improving this :smiley:

AdriVanHoudt commented:
:tada: :fireworks: no problem glad I could help! Are you open for a PR to do the error logging automagically in hapi?
